src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java

```diff
@@ -0,0 +1,1175 @@
+                } else {
+                    throw new CancellationException();
+                }
+            }
+        }
+
+        /**
+         * Throws the exception produced by the given exception supplying function if
+         * a task completed abnormally. If any task completed with an exception then
+         * the function is invoked with the exception of the first task to fail.
+         * If no task completed with an exception but cancelled tasks were notified to
+         * the {@code handleComplete} method then the function is called with a {@code
+         * CancellationException}. The exception returned by the function is thrown.
+         * This method does nothing if no tasks completed abnormally.
+         *
+         * <p> This method is intended to be invoked by the task scope owner after it has
+         * invoked {@link #join() join} (or {@link #joinUntil(Instant) joinUntil}).
+         * The behavior of this method is unspecified when invoking this method before
+         * the join is done.
+         *
+         * @param esf the exception supplying function
+         * @param <X> type of the exception to be thrown
+         * @throws X produced by the exception supplying function
+         */
+        public <X extends Throwable>
+        void throwIfFailed(Function<Throwable, ? extends X> esf) throws X {
+            Objects.requireNonNull(esf);
+            Future<Object> f = future;
+            if (f != null) {
+                Throwable throwable = (f.state() == Future.State.FAILED)
+                        ? f.exceptionNow()
+                        : new CancellationException();
+                X ex = esf.apply(throwable);
+                Objects.requireNonNull(ex, "esf returned null");
+                throw ex;
+            }
+        }
+    }
+
+    /**
+     * Holder class for the virtual thread factory. It uses reflection to allow
+     * this class be compiled in an incubator module without also enabling preview
+     * features.
+     */
+    private static class FactoryHolder {
+        static final ThreadFactory VIRTUAL_THREAD_FACTORY = virtualThreadFactory();
+
+        @SuppressWarnings("removal")
+        private static ThreadFactory virtualThreadFactory() {
+            PrivilegedAction<ThreadFactory> pa = () -> {
+                try {
+                    Method ofVirtualMethod = Thread.class.getDeclaredMethod("ofVirtual");
+                    Object virtualThreadBuilder = ofVirtualMethod.invoke(null);
+                    Class<?> ofVirtualClass = Class.forName("java.lang.Thread$Builder$OfVirtual");
+                    Method factoryMethod = ofVirtualClass.getMethod("factory");
+                    return (ThreadFactory) factoryMethod.invoke(virtualThreadBuilder);
+                } catch (Exception e) {
+                    throw new InternalError(e);
+                }
+            };
+            return AccessController.doPrivileged(pa);
```

Comment:
It might be better to use `MethodHandle`s obtained using <code>[jdk.internal.access.SharedSecrets]&#x200B;.getJavaLangInvokeAccess()</code> and <code>[JavaLangInvokeAccess]&#x200B;.findStatic(…)</code> and <code>[JavaLangInvokeAccess]&#x200B;.findVirtual(…)</code> for this, which would avoid going through `AccessController.doPrivilaged(…)`.

[jdk.internal.access.SharedSecrets]: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/access/SharedSecrets.java
[JavaLangInvokeAccess]: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/access/JavaLangInvokeAccess.java

Reply:
I'd prefer not export jdk.internal.access to this module, if possible.  In general, it's a lot easier to reason about the security and integrity of the core platform when java.base doesn't export any of its internal packages.

In any case, I expect this issue will resolve itself once there is a way for code in "core" modules to use preview APIs without needing to be compiled with --enable-preview. The java.management and jdk.incubator.vector modules have the same issue.